driver/src/contracts/stablex_contract.rs

```diff
@@ -59,14 +59,77 @@ impl StableXContract for StableXContractImpl {
 
     fn submit_solution(
         &self,
-        _batch_index: U256,
-        _orders: Vec<Order>,
-        _solution: Solution,
+        batch_index: U256,
+        orders: Vec<Order>,
+        solution: Solution,
     ) -> Result<()> {
-        unimplemented!();
+        let account = self
+            .base
+            .account_with_sufficient_balance()
+            .ok_or("Not enough balance to send Txs")?;
+
+        let (owners, order_ids, volumes, prices, token_ids_for_price) =
+            parse_auction_results(orders, solution);
+
+        self.base
+            .contract
+            .call(
+                "submitSolution",
+                (
+                    batch_index,
+                    owners,
+                    order_ids,
+                    volumes,
+                    prices,
+                    token_ids_for_price,
+                ),
+                account,
+                Options::default(),
+            )
+            .wait()
+            .map_err(DriverError::from)
+            .map(|_| ())
     }
 }
 
+type ContractRecognizedAuctionResults = (Vec<H160>, Vec<U128>, Vec<U128>, Vec<U128>, Vec<U128>);
+
+fn parse_auction_results(
+    orders: Vec<Order>,
+    solution: Solution,
+) -> ContractRecognizedAuctionResults {
+    // Representing the solution's price vector more compactly as:
+    // sorted_touched_token_ids, non_zero_prices which are logically bound by index.
+    // Example solution.prices = [3, 0, 1] will be transformed into [0, 2], [3, 1]
+    let mut ordered_token_ids: Vec<U128> = vec![];
+    let mut prices: Vec<U128> = vec![];
+    for (token_id, price) in solution.prices.into_iter().enumerate() {
+        if price > 0 {
+            ordered_token_ids.push(U128::from(token_id as usize));
+            prices.push(U128::from(price as usize));
+        }
+    }
+
+    let mut owners: Vec<H160> = vec![];
+    let mut order_ids: Vec<U128> = vec![];
+    let mut volumes: Vec<U128> = vec![];
+    let zipped_amounts = solution
+        .executed_buy_amounts
+        .into_iter()
+        .zip(solution.executed_sell_amounts.into_iter());
+    for (order_id, (buy_amount, sell_amount)) in zipped_amounts.enumerate() {
+        if buy_amount > 0 && sell_amount > 0 {
```

Comment:
Question for @josojo , @fleupold  and @twalth3r. I would like to confirm that these are the necessary and sufficient condition for an order to be considered "touched". I can imagine a situation of infinities where merely one is non-zero that would say buy zero of A for non-zero of B (or vice-versa). I imagine the contract would struggle with division errors or other such things and just wanna make sure we are all on the same page. 

I ask this, because I have unit test for each of the four cases: `(buyAmount, sellAmount) \in {zero, non-zero}^2`

Reply:
I believe that if the sellAmount is 0, we will also not process it in the smart contract. However, if the sellAmount is positive, it shouldn't matter what the buyVolum is, the smart contract (only receiving the sellVolume) would process it.

Therefore, I think the condition might only be

`if sell_amount  > 0`

Reply:
Good enough for me! Will readjust the unit test back to work for this and move on to the final issue regarding casting. Let us note that this condition only holds in the case that the order is a sell order.

```
U128::from(x as usize)
```
downcasts to u64 and will result in a loss of precision.